src/main/scala/chisel3/util/Decoupled.scala

```diff
@@ -215,28 +220,35 @@ class Queue[T <: Data](val gen: T,
     }
   }
 
-  val io = IO(new QueueIO(genType, entries))
-
+  val io = IO(new QueueIO(genType, entries, hasFlush))
   val ram = if (useSyncReadMem) SyncReadMem(entries, genType, SyncReadMem.WriteFirst) else Mem(entries, genType)
   val enq_ptr = Counter(entries)
   val deq_ptr = Counter(entries)
   val maybe_full = RegInit(false.B)
```

Comment:
Try to think about using logic below 
```
val maybe_full = RegNext(!flush && do_enq && do_deq)
```
to get rid of a Mux and a XORr gate.

Reply:
I don't think this suggestion is correct: it means that `maybe_full` cannot become true unless a dequeue occurs, which doesn't make sense.

Reply:
Oh, I thought about this, my bad!